### PR TITLE
Provide fix suggestions for not covered Maven plugin execution in project build lifecycle

### PR DESCRIPTION
**Pull request by testforstephen**
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

This PR requires the side PR at https://github.com/eclipse/eclipse.jdt.ls/pull/1770. It's used to mitigate the issues at https://github.com/redhat-developer/vscode-java/issues/177, https://github.com/redhat-developer/vscode-java/issues/566, https://github.com/redhat-developer/vscode-java/issues/935 and https://github.com/redhat-developer/vscode-java/issues/1910.

Here are the main things this PR added.
- Change the default value of preference "java.configuration.maven.notCoveredPluginExecutionSeverity" to "error", and show error diagnostics to user for awareness.
- Provide quickfix suggestions for not covered Maven plugin execution in pom.xml.

### CODE REVIEW

**Comment by testforstephen** [comment]
See a sample project https://github.com/antlr/stringtemplate4 to try this PR.
See a demo about the UX.

https://user-images.githubusercontent.com/14052197/118446113-d3ca8200-b721-11eb-8738-b11f6afe7f8c.mov

A follow-up improvement is to "discover new m2e connectors". Eclipse provides such a quick fix when reporting "not covered execution" error in Maven project. Looks like it's based on the catalog https://repo1.maven.org/maven2/.m2e/discovery-catalog/connectors.xml for discovery. Need more evaluation on the cost to support discovering and installing m2e connector, will take it as a future improvement.

// @fbricon Could you help review the solution? thanks.

File: document/_java.notCoveredExecution.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# Not Covered Maven Plugin Execution
+
+## Background
+Some Maven projects use 3rd party Maven plugins to generate sources or resources, and you may find that the generated code is not picked up by the project classpath, or the Maven goals are not executed by Java extension. This is a known technical issue with `m2e`, the underlying Maven integration tool for Java extension. The reason is that the Maven tooling `m2e` doesn't know if it's safe to run your Maven plugin automatically during a workspace build, so it does not run them by default and requires explicit instructions on how to handle them. Learn more about this issue from the wiki about [Execution Not Covered](https://www.eclipse.org/m2e/documentation/m2e-execution-not-covered.html).
+
+## Workaround
+- Option 1: Use [build-helper-maven-plugin](http://www.mojohaus.org/build-helper-maven-plugin/usage.html) to explicitly add the unrecoginzed source folder to classpath.
```
Comment (fbricon) [inline document/_java.notCoveredExecution.md:7]: unrecognized

**Review by fbricon — CHANGES_REQUESTED** [review]
I guess it will help in some cases, but I'm afraid we're opening a can of worms with the whole m2e connector thing (especially the discovery mechanism).

The best thing is still for maven plugins authors to provide [native support for m2e](https://www.eclipse.org/m2e/documentation/m2e-making-maven-plugins-compat.html). The doc should probably encourage users to request such integration directly from these projects.

**Comment by testforstephen** [comment]
Yes, the m2e connectors are a bit too complex and not planned to add them to VS Code. We are currently just providing some suggestions for common workarounds, and also stating in the doc that the best approach would be to make the Maven plugin directly compatible with m2e.

**Comment by testforstephen** [comment]
Update: 
- The latest commit has changed the default severity level to "warning".
- The PI didn't break the deployment to maven central.
Tried deploying a pom with PI `<?m2e execute onConfiguration?>` to oss.sonatype.org, it worked. And reference the sample dependency in other maven project, no issues were found.

**Comment by rgrunber** [comment]
I will look at this today.

File: package.json [hunk 2]
```diff
@@ -31,7 +31,8 @@
     "workspaceContains:pom.xml",
     "workspaceContains:build.gradle",
     "workspaceContains:.classpath",
-    "onCommand:_java.templateVariables"
+    "onCommand:_java.templateVariables",
+    "onCommand:_java.notCoveredExecution"
```
Comment (rgrunber) [inline package.json:35]: Is there a reason to add this ? From what I can tell, the command is only registered when the language server is activated.

**Review by rgrunber — CHANGES_REQUESTED** [review]
Seems to be working well. Just a few corner cases to consider.

File: src/pom/pomCodeActionProvider.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+				const indentation = this.getNewTextIndentation(document, diagnostic);
+				const saveAndUpdateConfigCommand: Command = {
+					title: "Save and update project configuration",
+					command: "_java.projectConfiguration.saveAndUpdate",
+					arguments: [ document.uri ],
+				};
+
+				const action1 = new CodeAction("Enable this execution in project configuration phase", CodeActionKind.QuickFix.append("pom"));
+				action1.edit = new WorkspaceEdit();
+				action1.edit.insert(document.uri, new Position(diagnostic.range.end.line, diagnostic.range.end.character + 1),
```
[the remark this answers is not in the document]
Comment (testforstephen) [reply src/pom/pomCodeActionProvider.ts]: Nice catch. Fixed.

File: src/pom/pomCodeActionProvider.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+	private getNewTextIndentation(document: TextDocument, diagnostic: Diagnostic): string {
+		const textline = document.lineAt(diagnostic.range.end.line);
+		if (textline.text.lastIndexOf("</execution>") > diagnostic.range.end.character) {
+			return "";
+		}
+
+		const lineSeparator = document.eol === EndOfLine.LF ? "\r" : "\r\n";
+		const currentTextIndentation = textline.text.substring(0, textline.firstNonWhitespaceCharacterIndex);
+		const tabSize = currentTextIndentation.length / 5;
```
[the remark this answers is not in the document]
Comment (testforstephen) [reply src/pom/pomCodeActionProvider.ts]: yes, using the tabSize from the active editor is more convenient.

File: package.json (hunk 2 above)
Comment (testforstephen) [reply package.json]: The quick fix is the only entry to open this document. Yes, no need to add it here.

**Review by rgrunber — APPROVED** [review]
LGTM. Feel free to merge when ready. I would probably squash the commits first.

**Pull request merged**